remme/ws/events.py

```diff
@@ -0,0 +1,168 @@
+            if entity['type'] == AtomicSwapInfo.__name__:
+                data = entity
+    if not data:
+        data = entities_changed
+    return transaction_id, data
+
+
+class WSEventSocketHandler(BasicWebSocketHandler):
+    def __init__(self, stream, loop):
+        super().__init__(stream, loop)
+        # events to subscribers
+        self._events = {event.value: [] for event in Events}
+        self._events_updator_task = weakref.ref(
+            asyncio.ensure_future(
+                self.listen_events(), loop=self._loop))
+        self.subscribe_events()
+
+    # return what value to be mapped to web_sock
+    async def subscribe(self, web_sock, entity, data):
+        if entity == Entity.EVENTS:
+            events = data.get('events', [])
+
+            if not events:
+                raise SocketException(web_sock, Status.EVENTS_NOT_PROVIDED, f"Events being subscribed are not provided")
+
+            for event in events:
+                if event not in self._events:
+                    raise SocketException(web_sock, Status.WRONG_EVENT_TYPE, f"Event: {event} is not supported")
+                if web_sock in self._events[event]:
+                    raise SocketException(web_sock, Status.ALREADY_SUBSCRIBED, f"Socket is already subscribed to: {event}")
+                self._events[event] += [web_sock]
+
+            LOGGER.info(f'Events being subscribed to: {events}')
+
+            return {'events': events}
+
+    def unsubscribe(self, web_sock):
+        # currently web_sock notification is based on self._subscriptions list
+        for event, web_socks in self._events.items():
+            if web_sock in web_socks:
+                web_socks.remove(web_sock)
+                self._events[event] = web_socks
+
+        pass
+
+    def subscribe_events(self):
+        # Setup a connection to the validator
+        LOGGER.info(f"Subscribing to events")
+        ctx = zmq.Context()
+        self._socket = ctx.socket(zmq.DEALER)
+        self._socket.connect(ZMQ_URL)
+        LOGGER.info(f"Connected to ZMQ")
+
+        request = ClientEventsSubscribeRequest(subscriptions=self._make_subscriptions(), last_known_block_ids=[]).SerializeToString()
+
+        # Construct the message wrapper
+        correlation_id = generate_random_key()  # This must be unique for all in-process requests
+        msg = Message(
+            correlation_id=correlation_id,
+            message_type=Message.CLIENT_EVENTS_SUBSCRIBE_REQUEST,
+            content=request)
+
+        # Send the request
+        LOGGER.info(f"Sending subscription request.")
+        self._socket.send_multipart([msg.SerializeToString()])
+
+        LOGGER.info(f"Subscribed.")
+
+    # The following code listens for events and logs them indefinitely.
+    async def check_event(self):
+        LOGGER.info(f"Checking for new events...")
+
+        resp = None
+        try:
+            resp = self._socket.recv_multipart(flags=zmq.NOBLOCK)[-1]
```

Comment:
Required to be async

Reply:
https://sawtooth.hyperledger.org/docs/core/releases/1.0/app_developers_guide/event_subscriptions.html#receiving-the-response

Reply:
It must be async in the event loop for preventing of holding a whole loop. Fast way to do - https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.AbstractEventLoop.run_in_executor

Reply:
With NOBLOCK flag it is not holding the loop
<img width="682" alt="screen shot 2018-08-15 at 3 48 57 pm" src="https://user-images.githubusercontent.com/6625496/44148940-e6c1785c-a0a2-11e8-9006-8f6ddf09d9cc.png">

Reply:
Yep, but it still sync call to socket and do not release a loop.

Reply:
check_event is purposefully triggered periodically. releasing the loop may create several async calls while waiting for a message and the function will execute several times for a single message.

Reply:
Ok, left it as it has, only wrap in executor `self._socket.send_multipart([msg.SerializeToString()])`.